middle_end/flambda2/simplify/expr_builder.ml

```diff
@@ -372,8 +372,13 @@ let create_let_symbol0 uacc (bound_symbols : Bound_symbols.t)
     if not will_bind_code
     then uacc
     else
+      let keep_code =
+        not
+          (Downwards_env.at_unit_toplevel
+             (Downwards_acc.denv (UA.creation_dacc uacc)))
```

Comment:
I'm reasonably sure that the `creation_dacc` is actually the `dacc` at the bottom of the downwards traversal (i.e. the one used at the creation of the `uacc`), so I don't think this does what you'd expect it to do: as it stands, `keep_code` will always be the same (because `creation_dacc` is constant), instead of depending on whether the expression to rebuild is at top_level or not.

Reply:
There is one traversal per function (plus one for the toplevel module), so in theory there are cases where `keep_code` is `true`. But you're right that in those cases, symbols are forbidden, so I probably need something more clever...